### ISSUE TITLE
[Feature] Allow setting of `cpuConcurrency` in config

### ISSUE THREAD

**Issue by pastelsky** [yarnpkg/berry#5635]
- [x] I'd be willing to implement this feature ([contributing guide](https://yarnpkg.com/advanced/contributing))
- [x] This feature is important to have in this repository; a contrib plugin wouldn't do

**Describe the user story**
Currently, yarn berry uses `os.availableParallelism` and falls back to `os.cpus` to decide concurrency for tasks it performs. While these might give the right values on most systems, certain CI systems (e.g. Bamboo) that use virtualization may not report the CPU count / affinity accurately since they count "milli CPUs" which can be a lot higher than the actual parallelism available. 

Here, setting an explicit CPU concurrency is needed, and this implicit assumption about CPUs fail. Most other tools in the JS ecosystem allow setting custom concurrency levels as well.

**Describe the solution you'd like**

Similar to the already available `networkConcurrency` option, consider adding a `cpuConcurrency` or `taskConcurrency` option that can be used to override concurrency in systems.

Allow this to be set at a command level — `yarn install --cpu-concurrency 5` or through an environment var — `YARN_CONCURRECNY=5 yarn install` so explicit values for the same can be set on CI.

**Describe the drawbacks of your solution**
Since this is an opt-in override, this shouldn't disrupt any current users.   

**Describe alternatives you've considered**
There are very few other alternatives we have, and none are particularly desirable —
- Forking yarn and patching its concurrency logic to use custom values from environment
- Patch node runtime before Yarn is run

**Comment by merceyz** [comment]
> Currently, yarn berry uses os.availableParallelism and falls back to os.cpus to decide concurrency for tasks it performs. While these might give the right values on most systems, certain CI systems (e.g. Bamboo) that use virtualization may not report the CPU count / affinity accurately since they count "milli CPUs" which can be a lot higher than the actual parallelism available.

That seems like something that should be reported to Node.js / libuv, maybe they can improve the function to take that into account.

**Comment by pastelsky** [comment]
> That seems like something that should be reported to Node.js / libuv, maybe they can improve the function to take that into account.

It isn't possible to statically determine such a limit. These CI systems have a shared CPU limit — which means a build can utilize more than what the machine might have. 

However, if all machines start over-utilizing, then there's global resource contention.

tldr; This needs to be set to a custom value depending on system load, machine type etc.
